performance/runner/src/calculate.rs

```diff
@@ -1,278 +1,169 @@
-                }
-            }
+        .map(|baseline| {
+            m_samples
+                .get(&baseline.metric)
+                .ok_or_else(|| RunnerError::BaselineMetricNotSampled(baseline.metric.clone()))
+                .map(|sample| calculate_regression(&sample, &baseline, 3.0))
         })
-        .collect::<Result<Vec<Vec<Calculation>>, CalculateError>>()?
-        .concat();
+        .collect::<Result<Vec<Calculation>, RunnerError>>()?;
 
     Ok(calculations)
 }
 
-// Top-level function. Given a path for the result directory, call the above
-// functions to compare and collect calculations. Calculations include both
-// metrics that fall within the threshold and regressions.
-pub fn regressions(results_directory: &PathBuf) -> Result<Vec<Calculation>, CalculateError> {
-    measurements_from_files(Path::new(&results_directory)).and_then(|v| {
-        // exit early with an Err if there are no results to process
-        if v.len() <= 0 {
-            Err(CalculateError::NoResultsErr(results_directory.clone()))
-        // we expect two runs for each project-metric pairing: one for each branch, baseline
-        // and dev. An odd result count is unexpected.
-        } else if v.len() % 2 == 1 {
-            Err(CalculateError::OddResultsCountErr(
-                v.len(),
-                results_directory.clone(),
-            ))
-        } else {
-            // otherwise, we can do our comparisons
-            let measurements = v
-                .iter()
-                // the way we're running these, the files will each contain exactly one measurement, hence `results[0]`
-                .map(|(p, ms)| (p, &ms.results[0]))
-                .collect::<Vec<(&PathBuf, &Measurement)>>();
-
-            calculate_regressions(&measurements[..])
-        }
-    })
-}
-
 #[cfg(test)]
 mod tests {
     use super::*;
+    use chrono::prelude::*;
 
     #[test]
-    fn detects_5_percent_regression() {
-        let dev = Measurement {
-            command: "some command".to_owned(),
-            mean: 1.06,
-            stddev: 1.06,
-            median: 1.06,
-            user: 1.06,
-            system: 1.06,
-            min: 1.06,
-            max: 1.06,
-            times: vec![],
+    fn detects_3sigma_regression() {
+        let metric = Metric {
+            name: "test".to_owned(),
+            project_name: "detects 3 sigma".to_owned(),
         };
 
-        let baseline = Measurement {
-            command: "some command".to_owned(),
-            mean: 1.00,
-            stddev: 1.00,
-            median: 1.00,
-            user: 1.00,
-            system: 1.00,
-            min: 1.00,
-            max: 1.00,
-            times: vec![],
+        let baseline = Baseline {
+            version: Version::new(9, 9, 9),
+            metric: metric.clone(),
+            ts: Utc::now(),
+            measurement: Measurement {
+                command: "some command".to_owned(),
```

Comment:
`to_owned` - what does this do? This is more of a curiosity question lol

Reply:
Literal strings in your Rust source code are references that live for the lifetime of your program (they can always be fetched from the binary). Because of all the lovely guarantees the rust typechecker makes for you around references, sometimes its necessary or just cleaner to own the string instead of borrowing it. `to_owned()` allocates the same sequence of utf-8 bytes on the heap behind a different pointer and lets you pass it around independently.

In this case when we make a value of type `Measurement` if I didn't use `to_owned` I would have to explicitly handle reference lifetimes in the code which is unnecessary complication for what we're doing here. Instead I just made a copy of the string and owned it to avoid that.

Reply:
This is pretty much the same thing every other language does, but you don't have the control to choose to use a pointer instead of a copy. In this case it probably looks like like unnecessary complication, but when you're doing anything more complicated than passing a string value it's a real life saver.

Reply:
Oh interesting. I see. `to_owned` is kinda a funny name but that making the copy instead of passing the same pointer around here makes sense